### PR TITLE
gmic,gmic-qt: 3.2.0 → 3.2.1, mark the latter as broken

### LINKED ISSUE #217082: Build failure: gmic
### Steps To Reproduce

1. build *gmic* from `master`

### Build log

```
/nix/store/cfdmgiycczhapzxgqg5m5y8a42lc7lak-cimg-3.2.1/include/CImg.h:32463:14: note:   candidate expects 7 arguments, 6 provided
/build/source/src/gmic.cpp: In instantiation of 'gmic_library::gmic_image<T>& gmic_library::gmic_image<T>::rol(const char*, gmic_library::gmic_list<T>&) [with T = float]':
/build/source/src/gmic.cpp:11742:9:   required from 'gmic& gmic::_run(const gmic_library::gmic_list<char>&, unsigned int&, gmic_library::gmic_list<T>&, gmic_library::gmic_list<char>&, gmic_library::gmic_list<T>&, >
/build/source/src/gmic.cpp:5406:14:   required from 'gmic& gmic::_run(const gmic_library::gmic_list<char>&, gmic_library::gmic_list<T>&, gmic_library::gmic_list<char>&) [with T = float]'
/build/source/src/gmic.cpp:4483:9:   required from 'gmic& gmic::_gmic(const char*, gmic_library::gmic_list<T>&, gmic_library::gmic_list<char>&, const char*, bool, float*, bool*) [with T = float]'
/build/source/src/gmic.cpp:2901:15:   required from here
/build/source/src/gmic.cpp:1767:28: error: no matching function for call to 'gmic_library::gmic_image<float>::_fill(const char* const&, bool, int, gmic_library::gmic_list<float>*, const char [4], gmic_library::gmi>
 1767 |   return rol((+*this)._fill(expression,true,1,&images,"rol",this));
      |              ~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/nix/store/cfdmgiycczhapzxgqg5m5y8a42lc7lak-cimg-3.2.1/include/CImg.h:32463:14: note: candidate: 'gmic_library::gmic_image<T>& gmic_library::gmic_image<T>::_fill(const char*, bool, unsigned int, gmic_library::gmic>
32463 |     CImg<T>& _fill(const char *const expression, const bool repeat_values, const unsigned int mode,
      |              ^~~~~
/nix/store/cfdmgiycczhapzxgqg5m5y8a42lc7lak-cimg-3.2.1/include/CImg.h:32463:14: note:   candidate expects 7 arguments, 6 provided
/build/source/src/gmic.cpp: In instantiation of 'gmic_library::gmic_image<T>& gmic_library::gmic_image<T>::ror(const char*, gmic_library::gmic_list<T>&) [with T = float]':
/build/source/src/gmic.cpp:11755:9:   required from 'gmic& gmic::_run(const gmic_library::gmic_list<char>&, unsigned int&, gmic_library::gmic_list<T>&, gmic_library::gmic_list<char>&, gmic_library::gmic_list<T>&, >
/build/source/src/gmic.cpp:5406:14:   required from 'gmic& gmic::_run(const gmic_library::gmic_list<char>&, gmic_library::gmic_list<T>&, gmic_library::gmic_list<char>&) [with T = float]'
/build/source/src/gmic.cpp:4483:9:   required from 'gmic& gmic::_gmic(const char*, gmic_library::gmic_list<T>&, gmic_library::gmic_list<char>&, const char*, bool, float*, bool*) [with T = float]'
/build/source/src/gmic.cpp:2901:15:   required from here
/build/source/src/gmic.cpp:1771:28: error: no matching function for call to 'gmic_library::gmic_image<float>::_fill(const char* const&, bool, int, gmic_library::gmic_list<float>*, const char [4], gmic_library::gmi>
 1771 |   return ror((+*this)._fill(expression,true,1,&images,"ror",this));
      |              ~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/nix/store/cfdmgiycczhapzxgqg5m5y8a42lc7lak-cimg-3.2.1/include/CImg.h:32463:14: note: candidate: 'gmic_library::gmic_image<T>& gmic_library::gmic_image<T>::_fill(const char*, bool, unsigned int, gmic_library::gmic>
32463 |     CImg<T>& _fill(const char *const expression, const bool repeat_values, const unsigned int mode,
      |              ^~~~~
/nix/store/cfdmgiycczhapzxgqg5m5y8a42lc7lak-cimg-3.2.1/include/CImg.h:32463:14: note:   candidate expects 7 arguments, 6 provided
/build/source/src/gmic.cpp: In instantiation of 'gmic_library::gmic_image<T>& gmic_library::gmic_image<T>::operator_minuseq(const char*, gmic_library::gmic_list<T>&) [with T = float]':
/build/source/src/gmic.cpp:12356:9:   required from 'gmic& gmic::_run(const gmic_library::gmic_list<char>&, unsigned int&, gmic_library::gmic_list<T>&, gmic_library::gmic_list<char>&, gmic_library::gmic_list<T>&, >
/build/source/src/gmic.cpp:5406:14:   required from 'gmic& gmic::_run(const gmic_library::gmic_list<char>&, gmic_library::gmic_list<T>&, gmic_library::gmic_list<char>&) [with T = float]'
/build/source/src/gmic.cpp:4483:9:   required from 'gmic& gmic::_gmic(const char*, gmic_library::gmic_list<T>&, gmic_library::gmic_list<char>&, const char*, bool, float*, bool*) [with T = float]'
/build/source/src/gmic.cpp:2901:15:   required from here
/build/source/src/gmic.cpp:1666:35: error: no matching function for call to 'gmic_library::gmic_image<float>::_fill(const char* const&, bool, int, gmic_library::gmic_list<float>*, const char [11], gmic_library::gm>
 1666 |   return operator-=((+*this)._fill(expression,true,1,&images,"operator-=",this));
      |                     ~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/nix/store/cfdmgiycczhapzxgqg5m5y8a42lc7lak-cimg-3.2.1/include/CImg.h:32463:14: note: candidate: 'gmic_library::gmic_image<T>& gmic_library::gmic_image<T>::_fill(const char*, bool, unsigned int, gmic_library::gmic>
32463 |     CImg<T>& _fill(const char *const expression, const bool repeat_values, const unsigned int mode,
      |              ^~~~~
/nix/store/cfdmgiycczhapzxgqg5m5y8a42lc7lak-cimg-3.2.1/include/CImg.h:32463:14: note:   candidate expects 7 arguments, 6 provided
/build/source/src/gmic.cpp: In instantiation of 'gmic_library::gmic_image<T>& gmic_library::gmic_image<T>::operator_xoreq(const char*, gmic_library::gmic_list<T>&) [with T = float]':
/build/source/src/gmic.cpp:13753:9:   required from 'gmic& gmic::_run(const gmic_library::gmic_list<char>&, unsigned int&, gmic_library::gmic_list<T>&, gmic_library::gmic_list<char>&, gmic_library::gmic_list<T>&, >
/build/source/src/gmic.cpp:5406:14:   required from 'gmic& gmic::_run(const gmic_library::gmic_list<char>&, gmic_library::gmic_list<T>&, gmic_library::gmic_list<char>&) [with T = float]'
/build/source/src/gmic.cpp:4483:9:   required from 'gmic& gmic::_gmic(const char*, gmic_library::gmic_list<T>&, gmic_library::gmic_list<char>&, const char*, bool, float*, bool*) [with T = float]'
/build/source/src/gmic.cpp:2901:15:   required from here
/build/source/src/gmic.cpp:1712:35: error: no matching function for call to 'gmic_library::gmic_image<float>::_fill(const char* const&, bool, int, gmic_library::gmic_list<float>*, const char [11], gmic_library::gm>
 1712 |   return operator^=((+*this)._fill(expression,true,1,&images,"operator^=",this));
      |                     ~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/nix/store/cfdmgiycczhapzxgqg5m5y8a42lc7lak-cimg-3.2.1/include/CImg.h:32463:14: note: candidate: 'gmic_library::gmic_image<T>& gmic_library::gmic_image<T>::_fill(const char*, bool, unsigned int, gmic_library::gmic>
32463 |     CImg<T>& _fill(const char *const expression, const bool repeat_values, const unsigned int mode,
      |              ^~~~~
/nix/store/cfdmgiycczhapzxgqg5m5y8a42lc7lak-cimg-3.2.1/include/CImg.h:32463:14: note:   candidate expects 7 arguments, 6 provided
```

### Additional context

There are more of those function arity errors, but just a small sample since they're the same. I'm going to assume #217015 pull in today has something to do with it.

### Notify maintainers

no maintainer? @AndersonTorres merged #217015 and it shows failures for gmic. gmic is used for me in GIMP and darktable.

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.12, NixOS, 23.05 (Stoat), 23.05.20230218.dirty`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.13.2`
 - channels(toastal): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

### PR DESCRIPTION
**Pull request by jtojnar**
###### Description of changes

Fixes #217082 

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by AndersonTorres — APPROVED** [review]

File: pkgs/tools/graphics/gmic-qt/default.nix [hunk 1]
```diff
@@ -97,7 +99,17 @@ mkDerivation rec {
     wrapQtApp "$out/${gimp.targetPluginDir}/gmic_gimp_qt/gmic_gimp_qt"
   '';
 
+  passthru = {
+    tests = {
+      gimp-plugin = gimpPlugins.gmic-qt;
```
Comment (lilyinstarlight) [inline pkgs/tools/graphics/gmic-qt/default.nix:104]:
```suggestion
      gimp-plugin = gimpPlugins.gmic;
```

I'm pretty sure it's just `gimpPlugins.gmic`, right?

**Review by lilyinstarlight — CHANGES_REQUESTED** [review]

File: pkgs/tools/graphics/gmic-qt/default.nix (hunk 1 above)
Comment (jtojnar) [reply pkgs/tools/graphics/gmic-qt/default.nix]: Good catch.

**Review by lilyinstarlight — APPROVED** [review]
This looks great to me now. Thank you @jtojnar!

**Review by wegank — APPROVED** [review]

**Comment by lilyinstarlight** [comment]
@jtojnar I had to undo their "fix" PR and make a gmic header indirection, but with this super hacky patch I can once again build gmic-qt with `ENABLE_SYSTEM_GMIC`:

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index ffd118c..3f2a790 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -290,7 +290,6 @@ if(ENABLE_CURL)
     endif()
 endif()
 
-add_definitions(-Dgmic_core)
 add_definitions(-Dgmic_community)
 add_definitions(-Dcimg_use_abort)
 add_definitions(-Dgmic_is_parallel)
@@ -298,6 +297,9 @@ add_definitions(-Dgmic_gui)
 add_definitions(-Dcimg_use_abort)
 add_definitions(-Dcimg_appname=\"gmic\")
 
+find_path(GMIC_INCLUDE_PATH gmic.h)
+add_definitions(-Dgmic_header=\"${GMIC_INCLUDE_PATH}/gmic.h\")
+
 if (UNIX)
     if(ANDROID)
         add_definitions(-Dcimg_display=0)
@@ -379,6 +381,7 @@ else()
 endif()
 
 if (NOT ENABLE_SYSTEM_GMIC)
+    add_definitions(-Dgmic_core)
     include_directories(${GMIC_PATH})
 endif (NOT ENABLE_SYSTEM_GMIC)
 include_directories(${CMAKE_SOURCE_DIR}/src)
diff --git a/src/gmic.h b/src/gmic.h
new file mode 100644
index 0000000..705fe2e
--- /dev/null
+++ b/src/gmic.h
@@ -0,0 +1,14 @@
+#ifndef gmic_version
+#ifndef gmic_core
+
+#define cimg_library gmic_library
+#define CImg gmic_image
+#define CImgList gmic_list
+
+#include "CImg.h"
+
+#endif // #ifndef gmic_core
+
+#include gmic_header
+
+#endif // #ifndef gmic_version
```

Basically it just includes `CImg.h` before `gmic.h` like it did before c-koi/gmic-qt#168 and redefines the CImg library namespace/classes like `gmic.h` does so that they use the new names

I may PR something upstream (not this hack but something more reasonable), because we are surely not the only ones bit by `ENABLE_SYSTEM_GMIC` being completely busted at the moment for gmic-qt

**Comment by lilyinstarlight** [comment]
@jtojnar Here is an upstream PR now that you can `fetchpatch` if you want: https://github.com/c-koi/gmic-qt/pull/174

**Comment by AndersonTorres** [comment]
I vote for pushing this and leaving the `fetchpatch` to the future.

**Comment by lilyinstarlight** [comment]
> I vote for pushing this and leaving the `fetchpatch` to the future.

Same. It shouldn't block this PR

**Pull request merged**